File: sudoq-app/sudoqmodel/src/test/java/de/sudoq/model/sudoku/CellTests.java
package de.sudoq.model.sudoku;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import org.junit.Test;

public class CellTests {

	@Test
	public void testNewCellWithSolution() {
		assertTrue(new Cell(false, 5, -1, 5).getCurrentValue() == 5);
	}

	@Test
	public void testSolution() {
		Cell f;

		try {
			f = new Cell(false, -2, -1, 3);
			fail("Initialisation with negative solution possible.");
		} catch (IllegalArgumentException e) {

		}

		for (int i = 0; i < 20; i += 5) {
			f = new Cell(false, i, -1, i);
			assertTrue("solutionFail", f.getSolution() == i);
		}
	}

	@Test
	public void testCurrent() {
		Cell f = new Cell(true, 8, -1, 20);

		try {
			f.setCurrentValue(Cell.EMPTYVAL - 1);
			fail("Initialisation with negative solution possible.");
		} catch (IllegalArgumentException e) {

		}

		try {
			f.setCurrentValue(Cell.EMPTYVAL - 1, false);
			fail("Initialisation with negative solution possible.");
		} catch (IllegalArgumentException e) {

		}

		for (int i = 0; i < 20; i += 5) {
			f.setCurrentValue(i);
			assertTrue("currentFail", f.getCurrentValue() == i);
		}
	}

	@Test
	public void testToggleNote() {
		Cell f = new Cell(-1, 1);

		for (int i = 0; i < 20; i++) {
			f.toggleNote(i);
			assertTrue("toggleN1 doesn't work", f.isNoteSet(i));
		}

		for (int i = 0; i < 20; i++) {
			f.toggleNote(i);
			assertTrue("toggleN2 doesn't work", !f.isNoteSet(i));
		}

		assertFalse("toggleN2 doesn't work", f.isNoteSet(-4));

	}

	@Test
	public void testEditable() {
		Cell f = new Cell(true, 6, -1, 9);
		assertTrue("editableFail", f.isEditable());
		assertTrue(f.getNumberOfValues() == 9);
		assertTrue(f.getId() == -1);
		f = new Cell(false, 6, -1, 9);
		assertFalse("editableFail", f.isEditable());

		f.setCurrentValue(3);
		assertTrue("editable doesn't lock", f.isNotWrong());
		f.setCurrentValue(-1);// darf nichts bewirken
		assertTrue("editable doesn't lock", f.isNotWrong());
	}

	@Test
	public void testIsSolvedCorrect() {
		Cell f = new Cell(true, 5, -1, 9);
		assertFalse("correctFail", f.isSolvedCorrect());
		f.setCurrentValue(5);
		assertTrue("correctFail", f.isSolvedCorrect());
		f.setCurrentValue(4);
		assertFalse("correctFail", f.isSolvedCorrect());
		f.setCurrentValue(5);
		assertTrue("correctFail", f.isSolvedCorrect());
	}

	@Test
	public void testIsNotWrong() {
		Cell f = new Cell(true, 4, -1, 9);
		f.setCurrentValue(0);
		assertFalse(f.isNotWrong());
	}

	@Test
	public void testEqual() {
		Cell f = new Cell(true, 3, -1, 9);
		Cell g = new Cell(true, 3, -1, 9);
		assertTrue(f.equals(g));

		g.toggleNote(2);

		assertFalse(f.equals(g));
		g.toggleNote(2);
		g.setCurrentValue(4);
		assertFalse(f.equals(g));

		assertFalse(f.equals(null));

		Cell h = new Cell(false, 2, -1, 9);
		assertFalse(f.equals(h));
	}

	@Test
	public void testClearIsEmpty() {
		Cell f = new Cell(true, 3, -1, 9);
		f.setCurrentValue(5);
		assertEquals(f.getCurrentValue(), 5);
		assertTrue(!f.isNotSolved());
		f.clearCurrentValue();
		assertEquals(f.getCurrentValue(), Cell.EMPTYVAL);
		assertTrue(f.isNotSolved());
	}

	@Test(expected = IllegalArgumentException.class)
	public void testSetCurrent() {
		Cell cell = new Cell(true, 0, 0, 9);
		cell.setCurrentValue(2, true);
		cell.setCurrentValue(-2, true);
	}

	@Test
	public void testToString() {
		Cell cell = new Cell(true, 0, 0, 21);
		cell.setCurrentValue(20);
		assertEquals("20", cell.toString());
	}

	@Test(expected = IllegalArgumentException.class)
	public void testValueTooHigh() {
		Cell cell = new Cell(true, 2, 0, 4);
		cell.setCurrentValue(4);
	}
}

File: sudoq-app/sudoqmodel/src/test/kotlin/de/sudoq/model/sudoku/CellTests.kt
package de.sudoq.model.sudoku

import org.amshove.kluent.*
import org.junit.jupiter.api.*
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class CellTests {


    @Test
    fun `should hold the solution that is passed`() {
        val cell = Cell(false, solution=5, -1, 5)
        cell.currentValue `should be equal to` 5
    }

    @Test
    fun `should not be possible to initialize with negative solution`() {
        invoking { Cell(false, solution=-2, -1, 3) } `should throw`
        IllegalArgumentException::class
    }

    @ParameterizedTest
    @ValueSource(ints = [0, 5, 10, 15])
    fun `should retain given solution`(solution: Int) {
        val f = Cell(false, solution, -1, solution)
        f.solution `should be equal to` solution
    }

    @Nested
    inner class SetCurrentValue {

        @Test
        fun `should not accept negative current value`() {
            val f = Cell(true, 8, -1, 20)
            invoking { run { f.currentValue = (Cell.EMPTYVAL - 1) } } `should throw`
                    IllegalArgumentException::class
        }

        @Test
        fun `should not accept negative current value with other method`() {
            val f = Cell(true, 8, -1, 20)
            invoking { f.setCurrentValue(Cell.EMPTYVAL - 1, false) } `should throw`
                    IllegalArgumentException::class
        }

        @ParameterizedTest
        @ValueSource(ints = [0,5,10,15])
        fun `should set value`(i: Int) {
            val f = Cell(true, 8, -1, 20)

            f.currentValue = i
			f.currentValue `should be equal to` i
		}
    }

    @Nested
    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
    @TestMethodOrder(MethodOrderer.OrderAnnotation::class)
    inner class ToggleNote {
        var f = Cell(-1, 1)

        @Order(1)
        @ParameterizedTest
        @ValueSource(ints = [0,1,2,3,4,5,6,7,8,9,10,12,13,14,15,16,17,18,19])
        fun `should toggle`(i: Int) {
            f.toggleNote(i)
            f.isNoteSet(i).`should be true`()
        }

        @Order(2)
        @ParameterizedTest
        @ValueSource(ints = [0,1,2,3,4,5,6,7,8,9,10,12,13,14,15,16,17,18,19])
        fun `should untoggle`(i: Int) {
            f.toggleNote(i)
            f.isNoteSet(i). `should be false`()
        }
    }

    @Test
    fun `should return the values it is initialized it with`() {
        val f = Cell(true, 6, -1, 9)
        f.isEditable.`should be true`()
        f.solution `should be equal to` 6
        f.id `should be equal to` (-1)
        f.numberOfValues `should be equal to` 9

        val g = Cell(false, 6, -1, 9)
        g.isEditable.`should be false`()
    }

    @Test
    fun `should not change when not editable`() {
        val f = Cell(false, 6, -1, 9)

        val initialValue = f.currentValue

        f.isEditable.`should be false`()

		f.currentValue = 3

        f.currentValue `should be` initialValue

        f.isNotWrong.`should be true`()

		f.currentValue = -1
        f.currentValue `should be` initialValue
        f.isNotWrong.`should be true`()
    }

    @Test
    fun testIsSolvedCorrect() {
        val f = Cell(true, 5, -1, 9)
        f.isSolvedCorrect.`should be false`()

        f.currentValue = 5
        f.isSolvedCorrect.`should be true`()

        f.currentValue = 4
        f.isSolvedCorrect.`should be false`()

        f.currentValue = 5
        f.isSolvedCorrect.`should be true`()
    }

    @Test
    fun testIsNotWrong() {
        val f = Cell(true, 4, -1, 9)
        f.currentValue = 0
        f.isNotWrong.`should be false`()
    }
    ///

    @Test
    fun testEqual() {
        val f = Cell(true, 3, -1, 9)
        val g = Cell(true, 3, -1, 9)

        f `should be equal to` g

        g.toggleNote(2)
        f `should not be equal to` g

        g.toggleNote(2)//revert
        f `should be equal to` g

        g.currentValue = 4
        f `should not be equal to` g
        f `should not be` null

        val h = Cell(false, 2, -1, 9)
        f `should not be equal to` h
    }

    @Test
    fun testClearIsEmpty() {
        val f = Cell(true, 3, -1, 9)
        f.currentValue = 5

        f.currentValue `should be` 5
        f.isNotSolved.`should be false`()

        f.clearCurrentValue()

        f.currentValue `should be` Cell.EMPTYVAL
        f.isNotSolved.`should be true`()
    }

    @Test
    fun `should throw when invoked with notify`() {
        val cell = Cell(true, 0, 0, 9)
        invoking { cell.setCurrentValue(-2, true) } `should throw` IllegalArgumentException::class
    }

    @Test
    fun testToString() {
        val cell = Cell(true, 0, 0, 21)
        cell.currentValue = 20
        cell.toString() `should be equal to` "20"
    }

    @Test
    open fun testValueTooHigh() {
        val cell = Cell(true, 2, 0, 4)
        invoking { run { cell.currentValue = 4 } } `should throw` IllegalArgumentException::class
    }
}